controller/sharding/sharding.go

```diff
@@ -460,21 +460,46 @@ func getDefaultShardMappingData(replicas int) []shardApplicationControllerMappin
 
 func GetClusterSharding(kubeClient kubernetes.Interface, settingsMgr *settings.SettingsManager, shardingAlgorithm string, enableDynamicClusterDistribution bool) (ClusterShardingCache, error) {
 	var replicasCount int
+
+	applicationControllerName := env.StringFromEnv(common.EnvAppControllerName, common.DefaultApplicationControllerName)
+	namespace := settingsMgr.GetNamespace()
+
 	if enableDynamicClusterDistribution {
-		applicationControllerName := env.StringFromEnv(common.EnvAppControllerName, common.DefaultApplicationControllerName)
-		appControllerDeployment, err := kubeClient.AppsV1().Deployments(settingsMgr.GetNamespace()).Get(context.Background(), applicationControllerName, metav1.GetOptions{})
+		appControllerDeployment, err := kubeClient.AppsV1().Deployments(namespace).Get(context.Background(), applicationControllerName, metav1.GetOptions{})
 		// if app controller deployment is not found when dynamic cluster distribution is enabled error out
 		if err != nil {
-			return nil, fmt.Errorf("(dynamic cluster distribution) failed to get app controller deployment: %w", err)
+			return nil, fmt.Errorf("(dynamic cluster distribution) failed to get app controller deployment '%s': %w", applicationControllerName, err)
 		}
 
-		if appControllerDeployment != nil && appControllerDeployment.Spec.Replicas != nil {
+		if appControllerDeployment.Spec.Replicas != nil {
 			replicasCount = int(*appControllerDeployment.Spec.Replicas)
 		} else {
-			return nil, fmt.Errorf("(dynamic cluster distribution) failed to get app controller deployment replica count")
+			return nil, fmt.Errorf("application controller deployment '%s' has no replicas specified", applicationControllerName)
 		}
 	} else {
-		replicasCount = env.ParseNumFromEnv(common.EnvControllerReplicas, 0, 0, math.MaxInt32)
+		// If dynamic cluster distribution is disabled, get the replica count from the StatefulSet.
+		appControllerStatefulSet, err := kubeClient.AppsV1().StatefulSets(namespace).Get(context.Background(), applicationControllerName, metav1.GetOptions{})
+		if err != nil {
+			replicasCount = 1
+			log.Warnf("Failed to retrieve StatefulSet '%s'. Defaulting replicasCount to 1.", applicationControllerName)
```

Comment:
As discussed in the issue, if the call fails, we should default to the environment variable. We cannot assume that the call will always work, and defaulting to 1 is not accurate.